### PR TITLE
Spell Checker

### PR DESCRIPTION
**Pull request by JPJPJPOPOP**
- Deconflicts #59
- Adds streamparser as a submodule

### CODE REVIEW

**Comment by sushain97** [comment]
We need to ignore streamparser. Modifying the Travis mypy command similarly should work: https://github.com/python/mypy/issues/626#issuecomment-279823106

File: modeSearch.py [hunk 1]
```diff
@@ -34,13 +34,17 @@ def searchPath(rootpath, include_pairs=True, verbosity=1):
         'pair': re.compile(r'({0})-({0})\.mode'.format(lang_code)),
         'analyzer': re.compile(r'(({0}(-{0})?)-(an)?mor(ph)?)\.mode'.format(lang_code)),
         'generator': re.compile(r'(({0}(-{0})?)-gener[A-z]*)\.mode'.format(lang_code)),
-        'tagger': re.compile(r'(({0}(-{0})?)-tagger)\.mode'.format(lang_code))
+        'tagger': re.compile(r'(({0}(-{0})?)-tagger)\.mode'.format(lang_code)),
+        'spell': re.compile(r'(({0}(-{0})?)-spell)\.mode'.format(lang_code)),
+        'tokenise': re.compile(r'(({0}(-{0})?)-tokenise)\.mode'.format(lang_code))
```
Comment (sushain97) [inline modeSearch.py]: add trailing comma

File: servlet.py [hunk 2]
```diff
@@ -908,6 +910,46 @@ def get(self):
...
+                    suggestion = []
+                    commands = [['apertium', '-d', path, '-f', formatting, mode]]
+
+                    result = yield translation.translateSimple(token.wordform, commands)
+                    foundSugg = False
+                    for line in result.split('\n'):
+                        if line.count('Corrections for'):
+                            foundSugg = True
+                            continue
+                        if foundSugg and '    ' in line:
```
Comment (sushain97) [inline servlet.py]: In think it's much clear if this and below are `\t`

File: servlet.py [hunk 3]
```diff
@@ -908,6 +910,46 @@ def get(self):
...
+            for token in tokens:
+                if token.knownness == known:
+                    units.append({'token': token.wordform, 'known': True, 'sugg': []})
+                else:
+                    suggestion = []
+                    commands = [['apertium', '-d', path, '-f', formatting, mode]]
+
+                    result = yield translation.translateSimple(token.wordform, commands)
+                    foundSugg = False
+                    for line in result.split('\n'):
```
Comment (sushain97) [inline servlet.py]: `splitlines`?

File: servlet.py [hunk 4]
```diff
@@ -908,6 +910,46 @@ def get(self):
...
+                            continue
+                        if foundSugg and '    ' in line:
+                            s, w = line.split('    ')
+                            suggestion.append((s, w))
+
+                    units.append({'token': token.wordform, 'known': False, 'sugg': suggestion})
+
+            self.sendResponse(units)
+        else:
+            logging.info('Spellchecker not working')
```
Comment (sushain97) [inline servlet.py]: why is this bad?? they input a wrong pair, right?

File: servlet.py [hunk 5]
```diff
@@ -1440,7 +1485,8 @@ def apply_config(args, apySection):
         (r'/identifyLang', IdentifyLangHandler),
         (r'/getLocale', GetLocaleHandler),
         (r'/pipedebug', PipeDebugHandler),
-        (r'/suggest', SuggestionHandler)
+        (r'/suggest', SuggestionHandler),
+        (r'/speller', SpellerHandler)
```
Comment (JPJPJPOPOP) [inline servlet.py]: Add comma at end

**Comment by sushain97** [comment]
Can you merge in changes from master? I'd like to test this locally.

File: servlet.py [hunk 6]
```diff
@@ -908,6 +910,45 @@ def get(self):
...
+                            continue
+                        if foundSugg and '\t' in line:
+                            s, w = line.split('\t')
+                            suggestion.append((s, w))
+
+                    units.append({'token': token.wordform, 'known': False, 'sugg': suggestion})
+
+            self.sendResponse(units)
+        else:
+            self.send_error(404, explanation="{} on spellchecker mode: {}".format('Error 404', 'Mode not installed'))
```
Comment (sushain97) [inline servlet.py]: This is kind of an awful error message. "Spelling mode kaz" not installed" e.g. would be far better.

**Comment by sushain97** [comment]
Hm... something seems suspect.

```
[14718:14714 - 0:825] 10:27:38 [Sushain@Zircon.local:o +1] ~/Documents/apertium-html-tools
$ curl 'http://localhost:2737/speller?q=personas&lang=oci'
[]
[14718:14714 - 0:826] 10:27:59 [Sushain@Zircon.local:o +1] ~/Documents/apertium-html-tools
$ curl 'http://localhost:2737/speller?q=persoeeenas&lang=oci'
[]
```

**Comment by jonorthwash** [comment]
@sushain97, is there a speller mode for oci?

**Comment by sushain97** [comment]
Yes.

**Pull request merged**